### ISSUE TITLE
Bug in STL10 dataset for all method with KNN, and val_acc_1

### ISSUE THREAD

**Issue by trungpx** [vturrisi/solo-learn#198]
Hello,
It is good library, we used it for several our papers. However, it is not completeness. 

I found two issues of this library when run on STL10 dataset as follows: 

**1)** since label for trained set comes from "train+unlabelled" then it doese have proper labels when do evaluation, so if you use "--knn_eval" to run any method, it will instantly got error and stop right away. Fail to train anymore. You may test and confirm again just less than 1 minute.

**2)** Compare to the code of this solo-learn one month ago before the update to the current version, the val_acc_1 in STL10 for all methods are much slower than the initial version (e.g. BYOL at 150 epoch have val_acc_1=82% for the version last month, but now I run your code, it is about 50%). I tried to find but didn't see the bug of your code, maybe you may find it. 

So could you fix these two issues and provide the bash files of all method for STL10, too? 

Thanks,
PXT

**Comment by DonkeyShot21** [comment]
Hi, thanks for reporting these issues. Please, if you use the library in your papers, please cite our publication. 

1) this is reasonable as we didn't test it with STL. Probably an easy fix. Will investigate.

2) did you notice this behavior only on STL or on other datasets as well? 50% accuracy is very low. We reran CIFAR10 and ImageNet100 recently and got the same results as before, but it's been a long time we haven't run STL10.

**Comment by trungpx** [comment]
Thanks for your quick feedback. Yes, we already cited your work of solo-learn in all of our publications recently using your bib in the main page. Before used openmmlab library [OpenSelfsup](https://github.com/open-mmlab/OpenSelfSup) now I found this library is good enough.

I found it performed well on other datasets like CIFAR10, CIFAR100, ImageNet-100. But STL10 has this issue. I couldn't find where is the problem. 

As notice before, results for STL10 in the code version 1 month before is nice. Only issue in this current version for STL10.

**Comment by vturrisi** [comment]
Hi @trungpx, thanks for using the library. If you can share the links to your papers it would be good, since we will probably do a "Who's using" section soon. I'm checking the code right now and I'll get back to you very soon.

Also, can you share your bash file?

**Comment by trungpx** [comment]
Hi vturrisi,
Our works have just passed the reviews process, awaiting for a while to get an official decision (one month later). So that I will share you the link of papers. 

Here is bash file for simsiam.sh

> python3 ../../../main_pretrain.py \
    --dataset stl10 \
    --encoder resnet18 \
    --data_dir /datasets/ \
    --max_epochs 150 \
    --gpus 6 \
    --precision 16 \
    --optimizer sgd \
    --scheduler warmup_cosine \
    --lr 0.5 \
    --classifier_lr 0.1 \
    --weight_decay 1e-5 \
    --batch_size 256 \
    --num_workers 4 \
    --brightness 0.4 \
    --contrast 0.4 \
    --saturation 0.4 \
    --hue 0.1 \
    --gaussian_prob 0.0 0.0 \
    --num_crops_per_aug 1 1 \
    --zero_init_residual \
    --name simsiam \
    --project solo-learn-stl10 \
    --entity trungpx \
    --wandb \
    --save_checkpoint \
    --method simsiam \
    --proj_hidden_dim 2048 \
    --pred_hidden_dim 512 \
    --proj_output_dim 2048 \
    # --crop_size 32 \
    # --knn_eval \

**And here for BYOL:**

> python3 ../../../main_pretrain.py \
    --dataset stl10 \
    --encoder resnet18 \
    --data_dir datasets/ \
    --max_epochs 150 \
    --gpus 7 \
    --precision 16 \
    --optimizer sgd \
    --lars \
    --grad_clip_lars \
    --eta_lars 0.02 \
    --exclude_bias_n_norm \
    --scheduler warmup_cosine \
    --lr 1.0 \
    --classifier_lr 0.1 \
    --weight_decay 1e-5 \
    --batch_size 256 \
    --num_workers 4 \
    --brightness 0.4 \
    --contrast 0.4 \
    --saturation 0.2 \
    --hue 0.1 \
    --gaussian_prob 0.0 0.0 \
    --solarization_prob 0.0 0.2 \
    --num_crops_per_aug 1 1 \
    --name byol \
    --project solo-learn-stl10 \
    --entity trungpx \
    --wandb \
    --save_checkpoint \
    --method byol \
    --proj_output_dim 256 \
    --proj_hidden_dim 4096 \
    --pred_hidden_dim 4096 \
    --base_tau_momentum 0.99 \
    --final_tau_momentum 1.0 \
    --momentum_classifier \
    # --crop_size 32 \
    # --knn_eval \

**Comment by vturrisi** [comment]
Sure, thanks! We've changed the augmentations pipeline, so now we need to manually pass `crop_size` for all datasets. If you do like this, your crop size will be 224x224 (which I assume it will stretch the image a lot or something like this). I'm running barlow (copy pasted from cifar10 and changed crop_size to 96), this is what I have at the moment:
![image](https://user-images.githubusercontent.com/34046216/143843417-4bdb468e-e5bd-477a-9b04-9b50685dd7a0.png)
About memory usage, that's also the reason, since you are forwarding a 224x224 image instead of a 96x96.

We will fix the knn stuff soon enough.

**Comment by trungpx** [comment]
Nice, thanks for hard works, I have fixed the manual crop_size to 96 as suggested, the issue of memory is gone now. Since previous it defined the pipeline of augmentation with crop_size for each dataset including STL10, there is no problem when remove --crop_size. But now it is, that the point as you have shown. 

I am now can fix them all. Will get back to you in the future.

**Comment by vturrisi** [comment]
Yes, we changed this behaviour in order to be more flexible with multicropping and the many different augmentation pipelines that are popping up.

**Comment by vturrisi** [comment]
Hey, in the latest version knn is now fixed. Also, when you update, don't forget to update library versions (we are now incompatible with older pytorch lightning versions).
